batchspawner/batchspawner.py

```diff
@@ -775,4 +777,106 @@ def state_gethost(self):
         self.log.error("Spawner unable to match host addr in job {0} with status {1}".format(self.job_id, self.job_status))
         return
 
+
+class OARSpawner(UserEnvMixin, BatchSpawnerBase):
+    batch_script = Unicode("""#!/bin/sh
+#OAR -n jupyterhub-singleuser-{{username}}
+#OAR -l nodes=1{% if nprocs     %}/core={{nprocs}}{% endif %}{% if runtime    %},walltime={{runtime}}{% endif %}
+{% if queue      %}#OAR -q {{queue}}{% endif %}
+{% if memory     %}#OAR -p "memnode>={{memory}}"{% endif %}
+{% if options    %}#OAR {{options}}{% endif %}
+
+{{prologue}}
+{{cmd}}
+{{epilogue}}
+""").tag(config=True)
+
+    # use a script to submit to batch queueing system instead of piping it in
+    script_mode = stat.S_IRUSR | stat.S_IWUSR | stat.S_IRGRP | stat.S_IROTH | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH #755 by default
+
+    script_home = os.path.join(os.environ.get("HOME", "/tmp"), "oar_jobs")
+    if not os.path.exists(script_home):
+        os.makedirs(script_home)
+    # outputs job id string
+    batch_submit_cmd = Unicode('/usr/bin/oarsub -d {homedir} -S ').tag(config=True)
+    # outputs job data XML string
+    batch_query_cmd = Unicode('/usr/bin/oarstat -J -j {job_id}').tag(config=True)
+    batch_cancel_cmd = Unicode('/usr/bin/oardel {job_id}').tag(config=True)
+
+    @gen.coroutine
+    def run_command(self, cmd, input=None, env=None):
```

Comment:
Copying run_command is probably the thing I would most want to update.  It's mostly the same, but saves the script file before submitting.  My basic suggestion is to override it, have it save the script file and prepare the new cmd, but then call the base class `run_command` to actually run everything.  Here's some pseudocode:

```
def run_command(self, cmd, input=None, env=None):
    ... Save the script file
    ... modify `cmd` to include that script file
    return super().run_command(cmd, input=None, env=env)
```
... of course modify the above to be a coroutine and all.

You may want to wait until after some of the upcoming changes go in, because it will drop compatibility with Python3.4 which might make things a bit easier.

We could also ask "why not directly add this to `run_command`".  That's a good question I can't answer, but I'd prefer to not make `run_command` more complicated unless another scheduler needs it, so we can see how robust it needs to be.  The above is basically the same thing, and can easily be moved to `run_command` when it's time.

Reply:
Dropping python3.4 may be a good point for you but here we are using OAR on debian8 (hence python3.4) and will drop both Debian8, OAR and Python3.4 at the same time. This is also why I made something quick.

Reply:
I guess that leaves you in a bad state with respect to future batchspawner, but the old version pre-minimum-3.5 should keep working just fine fro the time being, it has no major known problems.  Without 3.5 we'd probably be basically stuck here anyway, so no major loss.